OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,220 @@
+    }
+
+    private byte[] query(URL url, ParcelableProxy proxy) throws QueryFailedException {
+        try {
+            Timber.d("fetching from Web Key Directory with: " + url + " proxy: " + proxy.getProxy());
+
+            Request request = new Request.Builder()
+                    .url(url)
+                    .build();
+
+            OkHttpClient client = OkHttpClientFactory.getClientPinnedIfAvailable(url, proxy.getProxy());
+            Response response = client.newCall(request).execute();
+
+            if (response.isSuccessful()) {
+                return response.body().bytes();
+            } else {
+                // probably a 404 indicating that the key does not exist
+                throw new QueryFailedException("key not found in Web Key Directory." +
+                        "response:" + response);
+            }
+
+        } catch (IOException e) {
+            Timber.e(e, "IOException at Web Key Directory key download");
+            throw new QueryFailedException("Cannot connect to Web Key Directory. "
+                    + "Check your Internet connection!"
+                    + (proxy.getProxy() == Proxy.NO_PROXY ? "" : " Using proxy " + proxy.getProxy()));
+        }
+    }
+
+    @Override
+    public void add(String armoredKey, ParcelableProxy proxy) throws AddKeyException {
+        throw new UnsupportedOperationException("Uploading keys to Web Key Directory not supported");
+    }
+
+    /**
+     * Web Key Directory returns the entire key even during our searching phase.
+     *
+     * @throws UnsupportedOperationException if the key could not be parsed
+     */
+    @NonNull
+    private static ImportKeysListEntry getEntry(byte[] data)
+            throws UnsupportedOperationException, IOException, PgpGeneralException {
+        ImportKeysListEntry entry = new ImportKeysListEntry();
+        entry.setSecretKey(false); // keys imported from Web Key Directory are public
+
+        UncachedKeyRing ring = UncachedKeyRing.decodeFromData(data);
+        UncachedPublicKey key = ring.getPublicKey();
+
+        entry.setUserIds(key.getUnorderedUserIds());
+        entry.setPrimaryUserId(key.getPrimaryUserIdWithFallback());
+
+        entry.setKeyId(key.getKeyId());
+        entry.setParcelableKeyRing(ParcelableKeyRing.createFromEncodedBytes(data));
+        entry.setFingerprint(key.getFingerprint());
+
+        try {
+            entry.setBitStrength(key.getBitStrength());
+            final int algorithm = key.getAlgorithm();
+            entry.setAlgorithm(KeyFormattingUtils.getAlgorithmInfo(algorithm, key.getBitStrength(),
+                    key.getCurveOid()));
+        } catch (NumberFormatException | NullPointerException e) {
+            Timber.e(e, "Conversion for bit size, algorithm, or creation date failed.");
+            // can't use this key
+            throw new UnsupportedOperationException(
```

Comment:
`UnsupportedOperationException` is not the right exception to throw if something went wrong during parsing of the key. If it's a thing that shouldn't occur (e.g. you expect the to always be correct) it's an `AsseritonError`, if it's just malformed input data it should be some kind of `IOException` or perhaps `PgpGeneralException`